### DISCUSSION TITLE
waittocomplete depreciated - alternative?

### DISCUSSION THREAD

**Discussion by MasonGW** [TPC-Council/HammerDB#677] [Q&A] [answered]
Hello!
I've been working with / learning how to use HammerDB for just over a month now primarily using the CLI and writing my own tcl scripts. Recently I started to mess with waittocomplete (mentioned here: https://www.hammerdb.com/docs4.0/ch09s08.html and here: https://www.hammerdb.com/docs4.3/ch10s04.html), however when I attempted to use it in my script I received a message stating that waittocomplete has been depreciated. I have yet to find any documentation providing an alternative. If there is any documentation I have missed, please link it. Thanks for your help!

**Comment by sm-shaw** [comment, marked as answer]
Hi yes, `waittocomplete` is deprecated and no longer needed to be set manually.  Instead, this functionality is managed automatically (so the CLI is consistent with the GUI) and can be controlled with the keepalive_margin setting to determine how long the HammerDB CLI will wait after a workload has completed before terminating the virtual users. From v4.10 this is set to 60 seconds, prior to v4.10 it was 10 seconds, however you can set this to whatever you need if you have long-running operations such as a checkpoint after a workload has finished and don't want HammerDB to terminate it earlier. 

```
hammerdb>print generic
Generic Dictionary Settings
...
commandline          {
 sqlite_db        = TMP
 keepalive_margin = 60
 jobsoutput       = JSON
 jobs_disable     = 0
}
...

hammerdb>giset commandline keepalive_margin 600
Changed commandline:keepalive_margin from 60 to 600 for generic

hammerdb>
```
There is a discussion on this topic here https://github.com/TPC-Council/HammerDB/discussions/622, but basically you don't need `waittocomplete ` now as it is built-in. 

Note that Tcl is used for multithreaded performance as described here https://www.hammerdb.com/blog/uncategorized/why-tcl-is-700-faster-than-python-for-database-benchmarking/.  However, the CLI itself can run in python mode if you prefer scripting in python https://www.hammerdb.com/blog/uncategorized/hammerdb-v4-6-new-features-pt1-python-cli-interface/ performance will be exactly the same with either front-end (as well as the GUI) as they all run _exactly_ the same workload against the target database.

**Reply by sm-shaw** [reply]
Also have a look at the example scripts for a template https://www.hammerdb.com/blog/uncategorized/hammerdb-v4-7-new-features-pt-2-example-cli-scripts/

**Reply by MasonGW** [reply]
Thank you for the quick response, this is very helpful!
